### DISCUSSION TITLE
umi4关于npm包声明问题

### DISCUSSION THREAD

**Discussion by hlskysong** [umijs/umi#8523] [Q&A] [answered]
项目中引入了一个私有npm包, 目前没有声明文件,

在项目根目录下的 `typings.d.ts` 中增加
```ts
// 项目生成
import '@umijs/max/typings';

// 新增
declare module 'xxxx';
```
不起效果, 报错信息如下:

```
扩大中的模块名称无效。模块“xxx”解析到位于“xxx”处的非类型化模块，其无法扩大。
```

在 `src/.umi/typings.d.ts` 中新增

`declare module 'xxxx';` 就生效了,

why ?

按说是不应该修改 `src/.umi/typings.d.ts` 这个文件的

**Comment by fz6m** [comment, marked as answer]
Create `src/global.d.ts` :

```ts
// src/global.d.ts

declare module 'xxxx'
```

**Reply by WLyKan** [reply]
文件名可以随意吧，叫 typings.d.ts 也行

**Reply by aikin-vip** [reply]
同样的问题。。没啥用啊

**Comment by kuake** [comment]
发现把跟目录下typings.d.ts里面的import '@umijs/max/typings'; 删了也可以了

**Comment by HarryLit** [comment]
import 应该放到声明 module 的内部
```typescript
// 新增
declare module 'xxxx' {
    // 项目生成
    import '@umijs/max/typings';
}
```
